java-frontend/src/main/java/org/sonar/java/se/FlowComputation.java

```diff
@@ -143,41 +143,59 @@ private ExecutionPath(@Nullable ExplodedGraph.Edge edge, PSet<ExplodedGraph.Edge
-      Symbol newTrackSymbol = addFlowFromLearnedAssociations(edge, flowBuilder);
+
+      List<JavaFileScannerContext.Location> laFlow = learnedAssociation(edge)
+        .map(la -> flowFromLearnedAssociation(la, edge.parent))
+        .orElse(ImmutableList.of());
+      flowBuilder.addAll(laFlow);
+
+      Symbol newTrackSymbol = newTrackedSymbol(edge);
+
       Stream<Constraint> constraintStream = addFlowFromLearnedConstraints(edge, flowBuilder);
       return new ExecutionPath(edge, visited.add(edge), newTrackSymbol, flowBuilder.build(), visitedAllParents(edge) || constraintStream.anyMatch(terminateTraversal));
     }
 
-    private boolean visitedAllParents(ExplodedGraph.Edge edge) {
-      return edge.parent.edges().stream().allMatch(visited::contains);
+    Optional<LearnedAssociation> learnedAssociation(ExplodedGraph.Edge edge) {
+      return edge.learnedAssociations().stream()
+        .filter(la -> la.symbol().equals(trackSymbol))
+        .findAny();
     }
 
-    boolean notVisited(ExplodedGraph.Edge e) {
-      return !visited.contains(e);
+    List<JavaFileScannerContext.Location> flowFromLearnedAssociation(LearnedAssociation learnedAssociation, ExplodedGraph.Node node) {
+      if (trackSymbol == null) {
+        return ImmutableList.of();
+      }
+      ImmutableList.Builder<JavaFileScannerContext.Location> flowBuilder = ImmutableList.builder();
+      for (Class<? extends Constraint> domain : domains) {
+        Constraint constraint = node.programState.getConstraint(learnedAssociation.symbolicValue(), domain);
+        if (constraint != null) {
+          String message = String.format("'%s' is assigned %s.", learnedAssociation.symbol().name(), constraint.valueAsString());
+          flowBuilder.add(new JavaFileScannerContext.Location(message, node.programPoint.syntaxTree()));
+        }
+      }
+      return flowBuilder.build();
     }
 
     @Nullable
-    private Symbol addFlowFromLearnedAssociations(ExplodedGraph.Edge edge, ImmutableList.Builder<JavaFileScannerContext.Location> flowBuilder) {
+    private Symbol newTrackedSymbol(ExplodedGraph.Edge edge) {
       if (trackSymbol == null) {
         return null;
       }
-      ExplodedGraph.Node parent = edge.parent;
-      Optional<LearnedAssociation> learnedAssociation = edge.learnedAssociations().stream()
-        .filter(la -> la.symbol().equals(trackSymbol))
-        .findFirst();
-      if (learnedAssociation.isPresent()) {
-        LearnedAssociation la = learnedAssociation.get();
-        for (Class<? extends Constraint> domain : domains) {
-          Constraint constraint = parent.programState.getConstraint(la.symbolicValue(), domain);
-          if (constraint != null) {
-            flowBuilder.add(location(parent, String.format("'%s' is assigned %s.", la.symbol().name(), constraint.valueAsString())));
-          }
-        }
-        return parent.programState.getLastEvaluated();
+      if (learnedAssociation(edge).isPresent()) {
```

Comment:
I do not find that more readable, because returned value depends only on presence of learned association, so the lambda inside the `map` would not have use its parameter.